modules/ethereum/src/lib.rs

```diff
@@ -742,17 +850,87 @@ pub(crate) mod tests {
 						header: initial_header,
 						total_difficulty: 0.into(),
 						next_validators_set_id: 0,
+						last_signal_block: None,
 					},
 				)]
 				.into_iter()
 				.collect(),
 				next_validators_set_id: 1,
-				validators_sets: vec![(0, (hash, initial_validators))].into_iter().collect(),
+				validators_sets: vec![(
+					0,
+					ValidatorsSet {
+						validators: initial_validators,
+						signal_block: None,
+						enact_block: hash,
+					},
+				)]
+				.into_iter()
+				.collect(),
 				validators_sets_rc: vec![(0, 1)].into_iter().collect(),
 				scheduled_changes: HashMap::new(),
 			}
 		}
 
+		pub(crate) fn insert(&mut self, header: Header) {
+			let hash = header.hash();
+			self.headers_by_number.entry(header.number).or_default().push(hash);
+			self.headers.insert(
+				hash,
+				StoredHeader {
+					submitter: None,
+					header,
+					total_difficulty: 0.into(),
+					next_validators_set_id: 0,
+					last_signal_block: None,
+				},
+			);
+		}
+
+		pub(crate) fn change_validators_set_at(
+			&mut self,
+			number: u64,
+			finalized_set: Vec<Address>,
+			signalled_set: Option<Vec<Address>>,
+		) {
+			let set_id = self.next_validators_set_id;
+			self.next_validators_set_id += 1;
+			self.validators_sets.insert(
+				set_id,
+				ValidatorsSet {
+					validators: finalized_set,
+					signal_block: None,
+					enact_block: self.headers_by_number[&0][0],
+				},
+			);
+
+			self.headers
+				.get_mut(&self.headers_by_number[&number][0])
```

Comment:
Why do we do that only for block `0` on the list? Is it guaranteed to be canonical? Shouldn't we do that for every element on that list?

Reply:
Because that's what I needed in tests - they are not about forks, they are about cases when there's change scheduled by ancestors (no matter if there are forks, or not).